`./x dist --stage 2 $tool` broken after #138224

Creating distribution tarballs of tools with `./x dist --stage 2 $tool` broke after #138224. We (Ferrocene) know it fails at least with rustfmt and clippy. From what I can see, it stopped building the stage1 library, causing that failure.

<details><summary>Build log before the PR got merged:</summary>

```
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.07s
Building stage0 library artifacts (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.05s
Building compiler artifacts (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.26s
Creating a sysroot for stage1 compiler (use `rustup toolchain link 'name' build/host/stage1`)
Building stage0 tool lld-wrapper (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.12s
Building stage1 library artifacts (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.03s
Building compiler artifacts (stage1 -> stage2, x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.24s
Building tool clippy-driver (stage1 -> stage2, x86_64-unknown-linux-gnu)
   Compiling clippy v0.1.87 (/home/pietro/r/github/rust-lang/rust/master/src/tools/clippy)
   Compiling clippy_lints v0.1.87 (/home/pietro/r/github/rust-lang/rust/master/src/tools/clippy/clippy_lints)
```

</details>


<details><summary>Build log after the PR got merged:</summary>

```
Building bootstrap
   Compiling bootstrap v0.0.0 (/home/pietro/r/github/rust-lang/rust/master/src/bootstrap)
    Finished `dev` profile [unoptimized] target(s) in 5.83s
Building stage0 library artifacts (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.10s
Building compiler artifacts (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.33s
Creating a sysroot for stage1 compiler (use `rustup toolchain link 'name' build/host/stage1`)
Building stage0 tool lld-wrapper (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized] target(s) in 0.20s
Building tool clippy-driver (stage1 -> stage2, x86_64-unknown-linux-gnu)
   Compiling proc-macro2 v1.0.93
   Compiling unicode-ident v1.0.14
   Compiling serde v1.0.217
   Compiling stable_deref_trait v1.2.0
   Compiling litemap v0.7.4
   Compiling writeable v0.5.5
   Compiling smallvec v1.13.2
   Compiling icu_locid_transform_data v1.5.0
   Compiling icu_properties_data v1.5.0
   Compiling utf8_iter v1.0.4
   Compiling equivalent v1.0.1
   Compiling write16 v1.0.0
   Compiling icu_normalizer_data v1.5.0
   Compiling utf16_iter v1.0.5
   Compiling hashbrown v0.15.2
   Compiling rustc_apfloat v0.2.2+llvm-462a31f5a5ab
error[E0463]: can't find crate for `std`
  |
  = note: the `x86_64-unknown-linux-gnu` target may not be installed
  = help: consider adding the standard library to the sysroot with `x build library --target x86_64-unknown-linux-gnu`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

For more information about this error, try `rustc --explain E0463`.
error[E0463]: can't find crate for `core`
  |
  = note: the `x86_64-unknown-linux-gnu` target may not be installed
  = help: consider adding the standard library to the sysroot with `x build library --target x86_64-unknown-linux-gnu`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

error: could not compile `proc-macro2` (build script) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `unicode-ident` (lib) due to 1 previous error
error: could not compile `serde` (build script) due to 1 previous error
error: could not compile `icu_locid_transform_data` (lib) due to 1 previous error
error: could not compile `litemap` (lib) due to 1 previous error
error: could not compile `stable_deref_trait` (lib) due to 1 previous error
error: could not compile `icu_properties_data` (lib) due to 1 previous error
error: could not compile `write16` (lib) due to 1 previous error
error: could not compile `equivalent` (lib) due to 1 previous error
error: could not compile `utf8_iter` (lib) due to 1 previous error
error: could not compile `writeable` (lib) due to 1 previous error
error: could not compile `smallvec` (lib) due to 1 previous error
error: could not compile `icu_normalizer_data` (lib) due to 1 previous error
error: could not compile `utf16_iter` (lib) due to 1 previous error
error: could not compile `rustc_apfloat` (build script) due to 1 previous error
error: could not compile `hashbrown` (lib) due to 1 previous error
Build completed unsuccessfully in 0:00:08
```
</details>

--- a/src/bootstrap/src/core/build_steps/dist.rs
+++ b/src/bootstrap/src/core/build_steps/dist.rs
@@ -421,13 +421,13 @@ impl Step for Rustc {
                 builder.install(&rustdoc, &image.join("bin"), FileType::Executable);
             }
 
+            let ra_proc_macro_srv_compiler =
+                builder.compiler_for(compiler.stage, builder.config.build, compiler.host);
+            builder.ensure(compile::Rustc::new(ra_proc_macro_srv_compiler, compiler.host));
+
             if let Some(ra_proc_macro_srv) = builder.ensure_if_default(
                 tool::RustAnalyzerProcMacroSrv {
-                    compiler: builder.compiler_for(
-                        compiler.stage,
-                        builder.config.build,
-                        compiler.host,
-                    ),
+                    compiler: ra_proc_macro_srv_compiler,
                     target: compiler.host,
                 },
                 builder.kind,
@@ -1177,6 +1177,8 @@ impl Step for Cargo {
         let compiler = self.compiler;
         let target = self.target;
 
+        builder.ensure(compile::Rustc::new(compiler, target));
+
         let cargo = builder.ensure(tool::Cargo { compiler, target });
         let src = builder.src.join("src/tools/cargo");
         let etc = src.join("src/etc");
@@ -1231,6 +1233,8 @@ impl Step for RustAnalyzer {
         let compiler = self.compiler;
         let target = self.target;
 
+        builder.ensure(compile::Rustc::new(compiler, target));
+
         let rust_analyzer = builder.ensure(tool::RustAnalyzer { compiler, target });
 
         let mut tarball = Tarball::new(builder, "rust-analyzer", &target.triple);
@@ -1272,6 +1276,8 @@ impl Step for Clippy {
     fn run(self, builder: &Builder<'_>) -> Option<GeneratedTarball> {
         let compiler = self.compiler;
         let target = self.target;
+
+        builder.ensure(compile::Rustc::new(compiler, target));
 
         // Prepare the image directory
         // We expect clippy to build, because we've exited this step above if tool
@@ -1323,8 +1329,11 @@ impl Step for Miri {
         if !builder.build.unstable_features() {
             return None;
         }
+
         let compiler = self.compiler;
         let target = self.target;
+
+        builder.ensure(compile::Rustc::new(compiler, target));
 
         let miri = builder.ensure(tool::Miri { compiler, target });
         let cargomiri = builder.ensure(tool::CargoMiri { compiler, target });
@@ -1461,6 +1470,8 @@ impl Step for Rustfmt {
     fn run(self, builder: &Builder<'_>) -> Option<GeneratedTarball> {
         let compiler = self.compiler;
         let target = self.target;
+
+        builder.ensure(compile::Rustc::new(compiler, target));
 
         let rustfmt = builder.ensure(tool::Rustfmt { compiler, target });
         let cargofmt = builder.ensure(tool::Cargofmt { compiler, target });
@@ -2326,6 +2337,8 @@ impl Step for LlvmBitcodeLinker {
     fn run(self, builder: &Builder<'_>) -> Option<GeneratedTarball> {
         let compiler = self.compiler;
         let target = self.target;
+
+        builder.ensure(compile::Rustc::new(compiler, target));
 
         let llbc_linker =
             builder.ensure(tool::LlvmBitcodeLinker { compiler, target, extra_features: vec![] });

--- a/src/bootstrap/src/core/builder/tests.rs
+++ b/src/bootstrap/src/core/builder/tests.rs
@@ -408,6 +408,7 @@ mod dist {
     use pretty_assertions::assert_eq;
 
     use super::{Config, TEST_TRIPLE_1, TEST_TRIPLE_2, TEST_TRIPLE_3, first, run_build};
+    use crate::Flags;
     use crate::core::builder::*;
 
     fn configure(host: &[&str], target: &[&str]) -> Config {
@@ -642,6 +643,37 @@ mod dist {
                 compile::Assemble { target_compiler: Compiler::new(1, a) },
                 compile::Assemble { target_compiler: Compiler::new(2, a) },
                 compile::Assemble { target_compiler: Compiler::new(2, b) },
+            ]
+        );
+    }
+
+    /// This also serves as an important regression test for <https://github.com/rust-lang/rust/issues/138123>
+    /// and <https://github.com/rust-lang/rust/issues/138004>.
+    #[test]
+    fn dist_all_cross() {
+        let cmd_args =
+            &["dist", "--stage", "2", "--dry-run", "--config=/does/not/exist"].map(str::to_owned);
+        let config_str = r#"
+            [rust]
+            channel = "nightly"
+
+            [build]
+            extended = true
+
+            build = "i686-unknown-haiku"
+            host = ["i686-unknown-netbsd"]
+            target = ["i686-unknown-netbsd"]
+        "#;
+        let config = Config::parse_inner(Flags::parse(cmd_args), |&_| toml::from_str(config_str));
+        let mut cache = run_build(&[], config);
+
+        // Stage 2 `compile::Rustc` should **NEVER** be cached here.
+        assert_eq!(
+            first(cache.all::<compile::Rustc>()),
+            &[
+                rustc!(TEST_TRIPLE_1 => TEST_TRIPLE_1, stage = 0),
+                rustc!(TEST_TRIPLE_1 => TEST_TRIPLE_1, stage = 1),
+                rustc!(TEST_TRIPLE_1 => TEST_TRIPLE_3, stage = 1),
             ]
         );
     }

